File: ios/TrailPlayground/VehicleArtwork.swift
import SwiftUI
import CoreGraphics
import TrailCore
import TrailUI
import TrailEffects

@MainActor struct VehicleArtwork {
    struct Layer {
        let path: CGPath, fill: CGColor, stroke: CGColor?, width: Double
    }
    let id: String, height: Double, anchor: TrailPoint, size: Double, headingWindow: Double, duration: Double
    let layers: [Layer]

    /// Paths/colors are prepared once. Every frame changes only the drawing transform.
    func draw(in context: CGContext, pose: TrailPose) {
        context.saveGState()
        context.translateBy(x: pose.point.x,y: pose.point.y)
        context.rotate(by: pose.headingRadians + .pi/2)
        context.scaleBy(x: size/height,y: size/height)
        context.translateBy(x: -anchor.x,y: -anchor.y)
        context.setLineJoin(.round)
        for layer in layers {
            context.setFillColor(layer.fill); context.addPath(layer.path); context.fillPath()
            if let stroke=layer.stroke {
                context.setStrokeColor(stroke); context.setLineWidth(layer.width)
                context.addPath(layer.path); context.strokePath()
            }
        }
        context.restoreGState()
    }
    func animation(_ motion: TrailMotionPreset) -> TrailAnimationSpec {
        let base=motion.animation(duration: .seconds(duration),repeats: true)
        return TrailAnimations.custom(TrailSampler { time in
            base.sampler.sample(at: TrailTime(progress: journeyProgress(time.progress),cycle: time.cycle,elapsedSeconds: time.elapsedSeconds))
        },duration: .seconds(duration),repeats: true,reducedMotion: base.reducedMotion)
    }

    static let fleet: [String:VehicleArtwork] = {
        struct File: Decodable { let vehicles: [Record] }
        struct Record: Decodable {
            struct Shape: Decodable { let path: [[Double]], fill: String, stroke: String?, width: Double? }
            let id: String, height: Double, anchor: [Double], size: Double, headingWindow: Double, duration: Double, layers: [Shape]
        }
        let url=Bundle.main.url(forResource: "vehicles",withExtension: "json")!
        let file=try! JSONDecoder().decode(File.self,from: Data(contentsOf: url))
        func color(_ hex: String) -> CGColor {
            let argb=UInt32(hex.dropFirst(),radix: 16)!
            let components=[CGFloat((argb>>16)&255)/255,CGFloat((argb>>8)&255)/255,CGFloat(argb&255)/255,CGFloat(argb>>24)/255]
            return CGColor(colorSpace: CGColorSpace(name: CGColorSpace.sRGB)!,components: components)!
        }
        return Dictionary(uniqueKeysWithValues: file.vehicles.map { record in
            let layers=record.layers.map { shape in
                let path=CGMutablePath()
                for command in shape.path {
                    func point(_ i: Int) -> CGPoint { CGPoint(x: command[i],y: command[i+1]) }
                    switch Int(command[0]) {
                    case 0: path.move(to: point(1))
                    case 1: path.addLine(to: point(1))
                    case 2: path.addCurve(to: point(5),control1: point(1),control2: point(3))
                    case 3: path.addQuadCurve(to: point(3),control: point(1))
                    case 4: path.closeSubpath()
                    default: preconditionFailure("Unknown vehicle path command")
                    }
                }
                return Layer(path: path,fill: color(shape.fill),stroke: shape.stroke.map(color),width: shape.width ?? 1)
            }
            return (record.id,VehicleArtwork(id: record.id,height: record.height,anchor: TrailPoint(record.anchor[0],record.anchor[1]),
                size: record.size,headingWindow: record.headingWindow,duration: record.duration,layers: layers))
        })
    }()
}

/// Same eased time drives the route head and the vehicle, with no independent animation clock.
func journeyProgress(_ p: Double) -> Double { min(1,max(0,p*p*p*(10+p*(-15+6*p)))) }

@MainActor struct JourneyVehicle: View {
    let path: TrailPath, artwork: VehicleArtwork, playback: TrailPlayback, reduced: Bool
    var body: some View {
        let frame=playback.frame(layer: playback.player.effect.layers.count-1)
        let fraction=reduced ? 1 : frame.head ?? journeyProgress(playback.progress)
        let pose=path.pose(at: fraction,headingWindow: artwork.headingWindow,direction: reduced ? .forward : frame.headDirection)
        Canvas { context,_ in
            if let pose { context.withCGContext { artwork.draw(in: $0,pose: pose) } }
        }
        .allowsHitTesting(false).accessibilityHidden(true)
    }
}

File: android/playground/src/main/kotlin/dev/trail/playground/examples/VehicleArtwork.kt
package dev.trail.playground.examples

import android.content.Context
import android.graphics.Canvas as NativeCanvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Bitmap
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.platform.LocalContext
import dev.trail.compose.TrailPlayback
import dev.trail.core.*
import dev.trail.effects.TrailMotionPreset
import org.json.JSONObject
import kotlin.math.PI
import kotlin.time.Duration.Companion.seconds

internal class VehicleArtwork(
    val id: String, val height: Double, val anchor: TrailPoint,
    val size: Double, val headingWindow: Double, val duration: Double,
    private val layers: List<VehicleLayer>,
) {
    fun bitmap(density: Double): Bitmap {
        val extent = (size * 1.8 * density).toInt().coerceAtLeast(1)
        return Bitmap.createBitmap(extent, extent, Bitmap.Config.ARGB_8888).also { bitmap ->
            draw(NativeCanvas(bitmap), TrailPose(TrailPoint(extent / density / 2, extent / density / 2), -PI/2, 0), density)
        }
    }
    /** Cached paths/paints; draw with a transform, without rebuilding artwork each frame. */
    fun draw(canvas: NativeCanvas, pose: TrailPose, density: Double = 1.0) {
        val saved=canvas.save()
        canvas.translate((pose.point.x*density).toFloat(),(pose.point.y*density).toFloat())
        canvas.rotate(Math.toDegrees(pose.headingRadians+PI/2).toFloat())
        val scale=(density*size/height).toFloat(); canvas.scale(scale,scale)
        canvas.translate(-anchor.x.toFloat(),-anchor.y.toFloat())
        for(layer in layers) {
            canvas.drawPath(layer.path,layer.fill)
            layer.stroke?.let { canvas.drawPath(layer.path,it) }
        }
        canvas.restoreToCount(saved)
    }
    fun animation(motion: TrailMotionPreset): TrailAnimationSpec {
        val base=motion.animation(duration.seconds,repeat=true)
        return TrailAnimations.custom(TrailAnimation { time ->
            base.sampler.sample(TrailTime(journeyProgress(time.progress),time.cycle,time.elapsedSeconds))
        },duration.seconds,repeat=true,reducedMotion=base.reducedMotion)
    }
}

/** Gentle departure/arrival; position and line head receive exactly the same eased clock. */
internal fun journeyProgress(p: Double): Double = (p*p*p*(10+p*(-15+6*p))).coerceIn(0.0,1.0)

internal data class VehicleLayer(val path: Path, val fill: Paint, val stroke: Paint?)

@Composable internal fun rememberVehicleArtwork(id: String): VehicleArtwork {
    val context=LocalContext.current
    return remember(context) { loadVehicleArtwork(context) }.getValue(id)
}

internal fun loadVehicleArtwork(context: Context): Map<String,VehicleArtwork> {
    val rows=context.assets.open("vehicles.json").bufferedReader().use { JSONObject(it.readText()) }.getJSONArray("vehicles")
    return (0 until rows.length()).associate { index ->
        val row=rows.getJSONObject(index); val layers=row.getJSONArray("layers")
        val prepared=(0 until layers.length()).map { i ->
            val layer=layers.getJSONObject(i); val commands=layer.getJSONArray("path")
            val path=Path()
            for(j in 0 until commands.length()) {
                val command=commands.getJSONArray(j)
                fun p(n: Int)=command.getDouble(n).toFloat()
                when(command.getInt(0)) {
                    0 -> path.moveTo(p(1),p(2)); 1 -> path.lineTo(p(1),p(2))
                    2 -> path.cubicTo(p(1),p(2),p(3),p(4),p(5),p(6))
                    3 -> path.quadTo(p(1),p(2),p(3),p(4)); 4 -> path.close()
                    else -> error("Unknown vehicle path command")
                }
            }
            fun paint(color: String, stroke: Boolean)=Paint(Paint.ANTI_ALIAS_FLAG).apply {
                this.color=android.graphics.Color.parseColor(color)
                style=if(stroke) Paint.Style.STROKE else Paint.Style.FILL
                strokeWidth=layer.optDouble("width",1.0).toFloat(); strokeJoin=Paint.Join.ROUND
            }
            VehicleLayer(path,paint(layer.getString("fill"),false),
                if(layer.has("stroke")) paint(layer.getString("stroke"),true) else null)
        }
        val anchor=row.getJSONArray("anchor")
        row.getString("id") to VehicleArtwork(row.getString("id"),row.getDouble("height"),
            TrailPoint(anchor.getDouble(0),anchor.getDouble(1)),row.getDouble("size"),
            row.getDouble("headingWindow"),row.getDouble("duration"),prepared)
    }
}

@Composable internal fun JourneyVehicle(
    path: TrailPath, artwork: VehicleArtwork, playback: TrailPlayback, layer: Int, reduced: Boolean,
) {
    Canvas(Modifier.fillMaxSize().testTag("journeyVehicle")) {
        // Observe time while drawing: vehicle movement does not recompose the map.
        val frame=playback.frame(layer)
        val fraction=if(reduced) 1.0 else frame.head ?: journeyProgress(playback.progress)
        val pose=path.poseAt(fraction,artwork.headingWindow,if(reduced) TrailDirection.Forward else frame.headDirection)
        if(pose!=null) drawIntoCanvas { artwork.draw(it.nativeCanvas,pose,density.toDouble()) }
    }
}
